Sources/SuperBuilder/ITKSuperBuilder.m

```diff
@@ -0,0 +1,348 @@
+        ERROR_AND_RETURN(SuperBuilderErrorCodeNoDynamicallyDispatchedMethodAvailable, msg)
+    }
+
+    // Add trampoline
+    let typeEncoding = method_getTypeEncoding(method);
+    let trampoline = ITKGetTrampolineForTypeEncoding(typeEncoding);
+    let methodAdded = class_addMethod(originalClass, selector, trampoline, typeEncoding);
+    if (!methodAdded) {
+        let msg = [NSString stringWithFormat:@"Failed to add method for selector %@ to class %@", NSStringFromSelector(selector), NSStringFromClass(originalClass)];
+        ERROR_AND_RETURN(SuperBuilderErrorCodeFailedToAddMethod, msg)
+    }
+    return methodAdded;
+}
+
+// Control if the trampoline should also push/pop the floating point registers.
+// This is slightly slower and not needed for our simple implementation
+// However, even if you just use memcpy, you will want to enable this.
+// We keep this enabled to be doubly safe.
+#define PROTECT_FLOATING_POINT_REGISTERS 1
+
+// One thread local per thread should be enough
+_Thread_local struct objc_super _threadSuperStorage;
+
+static BOOL ITKMethodIsSuperTrampoline(Method method) {
+    let methodIMP = method_getImplementation(method);
+    return methodIMP == msgSendSuperTrampoline || methodIMP == msgSendSuperStretTrampoline;
+}
+
+struct objc_super *ITKReturnThreadSuper(__unsafe_unretained id obj, SEL _cmd);
+struct objc_super *ITKReturnThreadSuper(__unsafe_unretained id obj, SEL _cmd) {
+    /**
+     Assume you have a class hierarchy made of four classes `Level1` <- `Level2` <- `Level3` <- `Level4`,
+     with `Level1` implementing a method called `-sayHello`, not implemented elsewhere in descendants classes.
+
+     If you use: `[SuperBuilder addSuperInstanceMethodToClass:Level2.class selector:@selector(sayHello) error:NULL];`
+     to inject a _dummy_ implementation at `Level2`, the following will happen:
+
+     - Calling `-[Level2 sayHello]` works. The trampoline is called, the `super_class ` is found to be `Level1`, and the `-sayHello` parent implementation is called.
+     - Calling `-[LevelN sayHello]` for any N > 2 ends in an infinite recursion. Since the `obj` passed to the trampoline is a descendant of `Level2`, `objc_msgSendSuper2` will of course call the injected implementation on `Level2`, which in turn will call itself with the same arguments, again and again.
+
+     This is fixed by walking up the hierarchy until we find the class implementing the method.
+
+     Looking at the method implementation we can also skip subsequent super calls.
+     */
+    Class clazz = object_getClass(obj);
+    Class superclazz = class_getSuperclass(clazz);
+    do {
+        let superclassMethod = class_getInstanceMethod(superclazz, _cmd);
+        let sameMethods = class_getInstanceMethod(clazz, _cmd) == superclassMethod;
+        if (!sameMethods && !ITKMethodIsSuperTrampoline(superclassMethod)) {
+            break;
+        }
+        clazz = superclazz;
+        superclazz = class_getSuperclass(clazz);
+    }while (1);
```

Comment:
This is not going to work. Let's consider the same situation as in the description above, but now `Level3` and `Level4` classes override the `-sayHello` method. 
Also, there is no way to determine if Method uses our trampoline or not, since Method can contain user-defined IMP which can call our trampoline IMP.
Seems like the only correct and efficient way to implement this is to use vm_remap() and PC-relative addressing in the same way as this does original `imp_implementationWithBlock`.

Reply:
You're right - this is not a situation the current logic supports. It would be a fun challenge to try how complex a solution with real trampolines work. Happy to take PRs! I add this limitation to the readme for now.